core/src/main/java/org/apache/cxf/common/util/ASMHelper.java

```diff
@@ -342,7 +342,10 @@ public ClassWriter createClassWriter() {
         TypeHelperClassLoader loader = getTypeHelperClassLoader(l);
         return loader.lookupDefinedClass(className);
     }
-
+    public static void addExternalClass(String className, ClassLoader l, Class<?> cls) {
```

Comment:
How these methods are supposed to be used?

Reply:
inside quarkiverse-cxf inside a a servlet 
https://github.com/quarkiverse/quarkiverse-cxf/blob/master/runtime/src/main/java/io/quarkiverse/cxf/CXFQuarkusServlet.java#L50
or 
https://github.com/quarkiverse/quarkiverse-cxf/blob/master/runtime/src/main/java/io/quarkiverse/cxf/CxfClientProducer.java#L19

I can add
ASMHelper.addExternalClass('io.quarkiverse.test.FruitFactory', Class.forName('io.quarkiverse.test.FruitFactory'));
here it is just for factory but I will provide the full list of generated class wrapperHelper (for request and response), packageInfo, customException with faultInfo field, NamespacePrefixMapper

Reply:
augmentor will produce class during build time and send class generated through recorder to runtime.

Reply:
Is it possible to not make it quarkus specific? Here is my proposal:

1. Make ASMHelper a SPI set as cxf bus extension
2. Provide current default impls
3. Have a a) generate proxy at build time mojo AND b) a load first proxy (before generating it) strategy (note that it makes asm useless and must not trigger asm classloading to work).

3 is close to https://github.com/apache/openwebbeans/blob/1f83552e74c7966fc1009a3a7417900e7ce42b32/webbeans-impl/src/main/java/org/apache/webbeans/service/ClassLoaderProxyService.java#L85 for runtime and https://github.com/apache/openwebbeans/blob/1f83552e74c7966fc1009a3a7417900e7ce42b32/webbeans-impl/src/main/java/org/apache/webbeans/service/ClassLoaderProxyService.java#L62 for build time (and dump the proxies as .class).

This way the code is not quarkus specific and benefits all applications in JVM and native mode.

Reply:
I do not get the point that this is quarkus specific 
1. indeed it is not quarkus specific but graal specific because can be used by spring boot too. Or any system which can provide class directly to avoid class generation. Anyway, It was the other proposition in the jira. But I found that less intrusive. The issue if I do so, I will have to do it for more class (wrapper helper, exception,...) which are generated. That s why I choose this solution.

2.  If I understand correctly, I create a class factory proxy static which provide creator which return Class<T> and if not in Map<String, Class<?>> . and each time, AsmHelper is used, I add this ProxyClassFactory in between. And in my code I just call ProxyClassFactory.getProxies().put(...);

Sorry, I do not know what you mean by "SPI set as cxf bus extension" ?
Do you mean, adding interface and implementation in bus-extensions.txt and adding interface for ASMHelper and inherit ASMHelper an implemtnation for each asm version in getASMClass ?
It seems a big refactoring and not related to the proxy need for quarkus ?
I agree that static class is not a good solution for ASMHelper and it is better to have factory pattern but I need a little help on that to have more explanation to respect the cxf pattern (bus-extensions.txt and factory). Do you have any document / guide ?

Reply:
@dufoli:

1. it is quarkus specific until you have the generator in cxf codebase (which is not the case in the pr) so it is quarkus specific ;). In other words, current PR is a quarkus extension point because it is feature incomplete for all other env and there is no real reason for that - don't get me wrong, I understand you got stucked and solve your issue but I think to integrate mainstream code it should go a bit further otherwise there is no benefit to get it in cxf land.
2. you actually don't need this map, you just do a classloader.loadClass (take care it is not the tccl but cxf classloader, can be in bus extensions). Only small thing to take care is to not do it at runtime but only deploy time but it should already be the case (or with a lazy init pattern but only once). In quarkus land you just generate the class at build time and put it in runtime classes (as other proxies).
3. the spi point is to avoid to have to handle ow2.asm, xbean.asm7, xbean.asm8, xbean.asm9 etc.
4. to not make it global it must be in cxf bus extensions IMHO (as ClassUnwrapper for example). Each time a class is generated, it has a bus not far so it should be easy to read from there.

hope it makes sense

Reply:
@rmannibucau  thanks for response ! ;-)
1. hmm, I think, I got your point. indeed, you mean that I need to use cxf parsing system to get serviceInfo and generate all classes without loading servlet or client stuff. The issue is that quarkus use yandex to parse classes and gizmo lib to generate code during build time. I am not sure, I can reuse the cxf system to parse wsdl, and annoted classe and generate serviceInfo in order to generate all classes. Is it doable ? Because I need to inject particular classWriter for that. 
Is it your point ? I have think to a such system at the beginning but it was a nightmare to understand how to use cxf for that. Because it is not a usual case.
2. So you mean that I inject my own classLoader which contain my classes in order that classLoader see the class I have generated ?
I do not get you point fully but I will try to code what I understand and I will let you review the code and give feedback.

Reply:
@dufoli

1. jandex or not does not change much things, at the end you must register the generated class in graal so a loadClass will cover both cases, only the generation phase will be different but CXF must get one impl for this feature to make sense in CXF itself IMHO - and it shouldnt be that hard to do a mojo generating proxies.
2. you don't need any custom classloader since classes are generated and dumped (physically or not) at build time in the target/classes of your project so you have them at runtime and just need to read them at run time.

Reply:
TODO:
For point 1)
- I create ASMHelper as interface with public 
- rename ASMHelper class to ASMHelperImpl and make it inherite of interface.
- create subclass of ASMHelperImpl  for each version of ASM ? (not sure of this part )
- create class Proxy implements ASMHelper which do nothing but just load class from classLoader. For this part it is not clear because className is send through a visite on ClassWriter. So it mean I need a NoOpClassWriter wich just store className ?
- add in bus-extenstions.txt a line:
org.apache.cxf.common.util.ASMHelper$TypeHelperClassLoader:org.apache.cxf.common.util.ASMHelper$:true

Reply:
just to know if I am on right direction... @rmannibucau can you just check what I provide.
But I need to understand some things which seems odd to me.

Reply:
Hmm,

Think it mixes two things, sorry if I was not clear about this point.
Until now we were generating proxies with asm, that was it.
Now we will be able to load proxies without using asm at all so we must split loading and generation IMO and only trigger generation if not existing so it means 2 independent services: GeneratedClassLoader maybe and ClassGenerator (backend by ASM).
Splitting it avoids to have the generation API in build time generation case.
It also means CXF does not need the proxy code anymore since asmX integration can now belong to another module (we can create a cxf-asm-compat module for migration purposes maybe and only keep ow2 integration code in cxf-core and register it only if asm is available and ClassGenerator not registered in the bus already).
About the registration: I wouldn't register it by default but use something like: if (not bus has ClassGenerator) tryLoadOw2AsmClassGenerator().

Hope it makes sense.
In all case I think it goes in the very right direction, thanks a lot for the effort!

Reply:
indeed best is just to provide me pseudoCode of usage you expect. Because, you are talking about SPI for ASMHelper and ClassLoader so I am bit lost? I think we need a factory for both (because I have static class)

Reply:
Idea is to split the way to get the class - findClass (Generated Class Loader or finder, name was not very great since ambiguous with ClassLoader) - and the way to generate a class (all asm impls).
All static calls must become deprecated imho.

Reply:
Ok So, I am on good way, but long move. Because so much here use static...

Reply:
I have move asmHelper to instance now. I will continue tomorrow...

Reply:
thanks for help ;-)

Reply:
@rmannibucau I have implement all for (it do not build because not migrate for other and I have issue of syntax validation.) 
1) do you know how maven can fix syntax validation automatically ? be cause process-sources report error but do not fix them. (import sort, ...)
2) if you are ok with this solution, I will continue with other classes.
BTW, Huge patch 1200 LOC and it is not finished ;-)

Reply:
is there any live chat like zulip, because we are both french and it can be easier to exchange about solution directly?

Reply:
@dufoli we have slack (the-asf), can need to request an invite (don't fully recall the details) but it should be quite straight forward to connect.
For syntax validation I'm generally doing it manually (don't think maven-checkstyle-plugin can autoformat sources).
About the rest I'd say ASMHelperImpl can maybe be N classes and drop the abstraction, wdyt? Also the generation is not only about namespace mapper think we have some other proxies no?
but overall, it looks good.

Reply:
and I need to know how I can inject to bus.GetExtension(org.apache.cxf.common.spi.ClassLoaderService.class) the 
org.apache.cxf.common.spi.ClassLoaderProxyService$LoadFirst instead of 
org.apache.cxf.common.spi.ClassLoaderProxyService$GenerateJustInTime
and after how I inject my class...
through bus-extensions.txt , if create bus-extensions.txt in my quarkus augmentor with
org.apache.cxf.common.spi.ClassLoaderProxyService:org.apache.cxf.common.spi.ClassLoaderProxyService$GenerateJustInTime: true
is it ok ?
or I have to use code to declare a change of extension ?

second question is how to use cxf to parse code and wsdl to generate class . I was thinking to use my own ASMHelper with a quarkus class writer .

Reply:
@dufoli: i tend to do programmatically or configuration (spring) when configuring cxf deployment. I didn't dig into your quarkus integration but if you create yourself your bus you can do it quite trivially, if you don't, you fallback on a default factory and the txt file is needed (i would recommend you to not rely on it since it will be easier to interact with cxf without but it works).

To generate you should trigger a cxf deployment (potentially on a local storage) and let the spy service capture the proxies, then you dump them in the graal classes singleton I think - here again, having the bus means you have the service instance so it is quite trivial to run the cxf deployment and read the captured instances ;).

Reply:
@rmannibucau I have added basic codes for it but it is complicated for something. 
ClassCreator class generate the JAXButils createNamespaceWrapper and it work but it work because we are in same module (core) but other classcreator are in dependencies... so if we switch to new system class creation need to be moved to core which seems a bad idea to me.

Reply:
@dufoli not sure i'm following, the goal is to hide ASM abstraction and replace it by a functional API so it shouldn't be an issue. Also for something jaxb specific you can use another SPI in the bus, sounds fine for me. They can be implemented in the same module (ex: cxf-asm, cxf-xbean-asm9 and so on) and instantiated at need only. wdyt?

Reply:
ok got it. So it mean that each class writer  (ASMHelper) will need a dedicated service provider with classloader on one side and class generator on other side. I get it.

Reply:
The classloader part can be shared, we don't have to split the code which is the same.
The proxy generation will likely be forked until we keep the kind of facade we have (but it has some pitfalls and is not tested).
Now if you prefer a single asm module for now it works too as a first step - your PR is not exactly on that point, it just implies a refactoring which enables that point ;). Once we have extensions in the bus we can refactor more easily in a second step too. The only key point is to split bytecode generation and loading in extensions.

Open point: will corba module be dropped (I see it uses asmhelper but if it is deleted soon no need to handle it and then asm becomes an xml thing it seems which can makes things easier).

Reply:
I have produce all ASMHelper spi. Need to fix format and test tomorrow.

Reply:
+1, small note: when you read an extension, maybe check it is null and if so throw an exception saying "missing extension X, you can add module m1 or m2" to ensure user is aware of what happens.

Reply:
basic null ref check... ;-)
My current issue, is that I miss one class JAXBExtensionHelper and it is full static... and no bus available to getExtension...
I am looking for solution.

Reply:
AFAIK it is only used in "register" or "loader" classes which have the bus ;)

Reply:
Yes I remove static...

Reply:
I have issue with test class. I use the mock system to create bus and return the good ASMHelper.

```
        Bus bus = control.createMock(Bus.class);
        EasyMock.expect(bus.getExtension(ASMHelper.class)).andReturn(new ASMHelperImpl()).anyTimes();
        ExtensionClassCreator extr = new ExtensionClassGenerator(bus);
        EasyMock.expect(bus.getExtension(ExtensionClassCreator.class)).andStubReturn(extr);
```

but I get error:
java.lang.IllegalStateException:
missing behavior definition for the preceding method call:
Bus.getExtension(interface org.apache.cxf.common.util.ASMHelper)
Usage is: expect(a.foo()).andXXX()
        at org.apache.cxf.wsdl.JAXBExtensionHelperTest.testAddTestExtension(JAXBExtensionHelperTest.java:88)

Reply:
I have try with AndStubReturn but change nothing...

Reply:
Before digging, did you try not mocking it at all? Looks very doable using the bus impl directly (also mock tests have very few value at the end so I wouldnt invest much in them if possible).

Reply:
it work without mock but now, I am on a real issue because asm version generated is not the good one ... Strange